### PR TITLE
[MM-12452] Stop inserting hashed plugin keys

### PR DESCRIPTION
**Pull request by gupsho**
#### Summary
Stop hashing keys for plugins when inserting new key/values. 

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/9813

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Added or updated unit tests (required for all new features)

### CODE REVIEW

File: app/plugin_test.go [hunk 1]
```diff
@@ -50,17 +50,15 @@ func TestPluginKeyValueStore(t *testing.T) {
 
 	// Test ListKeys
 	assert.Nil(t, th.App.SetPluginKey(pluginId, "key2", []byte("test")))
-	hashedKey := getHashedKey("key")
-	hashedKey2 := getHashedKey("key2")
 	list, err := th.App.ListPluginKeys(pluginId, 0, 1)
 	assert.Nil(t, err)
 	assert.Equal(t, 1, len(list))
-	assert.Equal(t, hashedKey, list[0])
+	assert.Equal(t, "key", list[0])
```
Comment (DSchalla) [inline app/plugin_test.go]: Can we add some tests for the fallback behavior, in case we have a hashed key already in the database and query it using the normal key, and the data is returned correct? We can enforce the old value creation using `th.App.Srv.Store.Plugin().SaveOrUpdate(kv)` in a test, with a KV struct where the key is already hashed and then query it afterwards. Similar, we should test the `Set` method replacing the old version.

File: app/plugin_key_value_store.go [hunk 2]
```diff
@@ -23,14 +23,13 @@ func (a *App) SetPluginKey(pluginId string, key string, value []byte) *model.App
...
 func (a *App) SetPluginKeyWithExpiry(pluginId string, key string, value []byte, expireInSeconds int64) *model.AppError {
-
 	if expireInSeconds > 0 {
 		expireInSeconds = model.GetMillis() + (expireInSeconds * 1000)
 	}
 
 	kv := &model.PluginKeyValue{
 		PluginId: pluginId,
-		Key:      getKeyHash(key),
+		Key:      key,
```
Comment (DSchalla) [inline app/plugin_key_value_store.go]: If we simply replace the set method using `key` instead of `getKeyHash(key)` we will end up with potentially duplicates and/or inconsistent data (from a plugin PoV) - We should use that function call to ensure that the hash version will be deleted (via `result := <-a.Srv.Store.Plugin().Delete(pluginId, getKeyHash(key))`) - That way we can slowly clean up the old K/V entries from the database until removing the transition code completely.

File: app/plugin_key_value_store.go [hunk 3]
```diff
@@ -45,18 +44,27 @@ func (a *App) SetPluginKeyWithExpiry(pluginId string, key string, value []byte,
...
+	}
 
+	result = <-a.Srv.Store.Plugin().Get(pluginId, getKeyHash(key))
 	if result.Err != nil {
 		if result.Err.StatusCode == http.StatusNotFound {
 			return nil, nil
 		}
 		mlog.Error(result.Err.Error())
 		return nil, result.Err
 	}
```
Comment (DSchalla) [inline app/plugin_key_value_store.go]: If we didn't return at this point yet, we should update the hashed key with the normal key in the database, to not only phase the hashed versions out via the `Set` method, but also via `Get`. Considering we got all information at line 67, we would only need to call effectively `SetPluginKeyWithExpiry`, copying all values from the function call / the returned KV Value from the database fetch. Combined with the requested change from the `SetPluginKeyWithExpiry` that will then 1. create the new version and 2. clean up the old value.

File: app/plugin_key_value_store.go [hunk 4]
```diff
@@ -45,8 +47,19 @@ func (a *App) SetPluginKeyWithExpiry(pluginId string, key string, value []byte,
 }
 
 func (a *App) GetPluginKey(pluginId string, key string) ([]byte, *model.AppError) {
-	result := <-a.Srv.Store.Plugin().Get(pluginId, getKeyHash(key))
+	result := <-a.Srv.Store.Plugin().Get(pluginId, key)
+
+	if result.Err == nil {
+		kv := result.Data.(*model.PluginKeyValue)
+		return kv.Value, nil
+	} else if result.Err != nil {
```
Comment (hanzei) [inline app/plugin_key_value_store.go]: Nit pick: You can drop the `else` clause here. If `result.Err == nil` the method returns any way.

Comment (gupsho) [reply app/plugin_key_value_store.go]: Done.

File: app/plugin_key_value_store.go [hunk 5]
```diff
@@ -57,6 +65,17 @@ func (a *App) GetPluginKey(pluginId string, key string) ([]byte, *model.AppError
 
 	kv := result.Data.(*model.PluginKeyValue)
 
+	// If we are here that means we are using old hashed key. Remove it and insert the new key without hashing it
+	_ = <-a.Srv.Store.Plugin().Delete(pluginId, getKeyHash(key))
+
+	// Insert the key without hashing
```
Comment (lieut-data) [inline app/plugin_key_value_store.go]: If we keep this code, I think we can just rely on `SetPluginKeyWithExpiry` deleting the old key, right?

File: app/plugin_key_value_store.go [hunk 6]
```diff
@@ -23,18 +23,20 @@ func (a *App) SetPluginKey(pluginId string, key string, value []byte) *model.App
...
 
 	kv := &model.PluginKeyValue{
 		PluginId: pluginId,
-		Key:      getKeyHash(key),
+		Key:      key,
 		Value:    value,
 		ExpireAt: expireInSeconds,
 	}
 
+	// First try deleting hashed key, then set using unhashed key
```
Comment (lieut-data) [inline app/plugin_key_value_store.go]: I'm inclined to think we should insert the new key and then try to delete the old one (logging but otherwise ignoring any error). This avoids data loss in the event of an extreme edge case.

File: app/plugin_key_value_store.go [hunk 7]
```diff
@@ -57,6 +65,17 @@ func (a *App) GetPluginKey(pluginId string, key string) ([]byte, *model.AppError
 
 	kv := result.Data.(*model.PluginKeyValue)
 
+	// If we are here that means we are using old hashed key. Remove it and insert the new key without hashing it
```
Comment (lieut-data) [inline app/plugin_key_value_store.go]: I'm inclined not to write to the key value storage on a get operation.

I assume the thinking here is to slowly migrate the keying strategy, but completing that assumes 1) all input keys get fetched in a timely fashion by "natural processes", and 2) we have some way of knowing that we're done converting keys. Since we can't guarantee the former (who knows how plugins use keys), and we can't detect the latter (no way to provably distinguish a hashed key from an unhashed one unless we add a column), I don't think there's a future where we can remove this code. In that case, is there any point in the migration?

File: app/plugin_test.go [hunk 8]
```diff
@@ -48,19 +48,39 @@ func TestPluginKeyValueStore(t *testing.T) {
...
+		PluginId: pluginId,
+		Key:      hashedKey2,
+		Value:    []byte("test"),
+		ExpireAt: 0,
+	}
+
+	result := <-th.App.Srv.Store.Plugin().SaveOrUpdate(kv)
+	assert.Nil(t, result.Err)
+
+	// Test fetch by hashed keyname
```
Comment (lieut-data) [inline app/plugin_test.go]: This comment and the one line line 67 seem like they should be swapped? This is testing the key, and the latter the hashed key, right?

Comment (lieut-data) [reply app/plugin_test.go]: Also, I can't comment on earlier lines, but:
* line 39: Can you add a check that verifies that `GetPluginKey` now returns `test2`?

File: app/plugin_test.go [hunk 9]
```diff
@@ -48,19 +48,39 @@ func TestPluginKeyValueStore(t *testing.T) {
...
+	ret, err = th.App.GetPluginKey(pluginId, "key2")
+	assert.Nil(t, err)
+	assert.Equal(t, kv.Value, ret)
+
+	// Test fetch by keyname
+	// The key should be changed to unhashed version because of earlier GetPluginKey, so we will get nil back
```
Comment (lieut-data) [inline app/plugin_test.go]: I don't understand this comment/test: shouldn't `hashedKey2` forever return the same result as fetching `"key2"`? Otherwise, if my plugin stores keys in an outside system, we'll lose the entry.

File: app/plugin_test.go [hunk 10]
```diff
@@ -82,6 +102,10 @@ func TestPluginKeyValueStore(t *testing.T) {
 	list, err = th.App.ListPluginKeys(pluginId, 2, 2)
 	assert.Nil(t, err)
 	assert.Equal(t, 0, len(list))
+
+	// Clean up
+	assert.Nil(t, th.App.DeletePluginKey(pluginId, "key"))
```
Comment (lieut-data) [inline app/plugin_test.go]: This is orthogonal to this change, but if a panic or `t.Fatal` occurs above, the test won't clean up, and we won't necessarily have a clean database. I'm addressing this kind of issue in a parallel change, but I would suggest at least moving this to the top of the function as a `defer()` statement.

File: app/plugin_key_value_store.go (hunk 7 above)
Comment (DSchalla) [reply app/plugin_key_value_store.go]: Considering plugins are still in beta, and we would update the key format in gets and sets, it would be reasonable to give a migration period of 2-3 releases and then remove the fallback code. Adding a temporary column wouldn't give us any advantage, considering even if we _know_ there are still some outdated keys in the DB, we cannot find out what's their key. Supporting hashed keys forever doesn't sound like a good approach for me.

What you think @jwilander @lieut-data ?

File: app/plugin_key_value_store.go (hunk 5 above)
Comment (DSchalla) [reply app/plugin_key_value_store.go]: Correct

File: app/plugin_key_value_store.go (hunk 7 above)
Comment (jwilander) [reply app/plugin_key_value_store.go]: What's the downside of keeping the code around forever? Having to maintain it and any misses on the regular key cause another get for the hashed key? That's not terrible. My worry about trying to migrate as Jesse says by "natural processes" is that in 2-3 releases when we make the breaking change we'll 1) be stable for plugins and 2) it could quietly cause issues for some plugins without any clear path for them to notice or fix the problem.

So if we're going to make a breaking change I think we should do it now and if we don't want to do that I think we need to keep the code around indefinitely unless it becomes a real problem for whatever reason

Comment (DSchalla) [reply app/plugin_key_value_store.go]: No, as you've said yourself it's neither a high performance impact, or is the code really hard to read that way. I was mostly questioning whether it's worth to break the current API (and essentially force a wipe of the KV Table with the next migration, since it's worthless data) to clean up the schema and the old hashed key format, or we are fine with keeping the code for future.
Personally not having a particular strong feeling about either option.

Comment (lieut-data) [reply app/plugin_key_value_store.go]: Forcing a wipe seems unusually painful -- e.g., this would force all the GitHub plugin users to re-sign in, for example.

With respect to the original question, another concern is that if you list the keys, then get a key, then list the keys again, the set will change, perhaps unexpectedly. It seems reasonable to do this on an `UPDATE` (changing the keys), but surprising otherwise.

File: app/plugin_test.go (hunk 9 above)
Comment (gupsho) [reply app/plugin_test.go]: HashedKey2 will be removed when we do a get for key2 because we are actively deleting hashed values to phase out old hashedKeys when a new set or get is done.

File: app/plugin_test.go (hunk 10 above)
Comment (gupsho) [reply app/plugin_test.go]: Done.

File: app/plugin_key_value_store.go (hunk 7 above)
Comment (gupsho) [reply app/plugin_key_value_store.go]: Make sense. Do we have a consensus to not update records on a get?

Comment (jwilander) [reply app/plugin_key_value_store.go]: I'm for not updating the keys

Comment (hanzei) [reply app/plugin_key_value_store.go]: One upside of updating the keys is that `KVList()` becomes more useful. It we never update the keys plugins will never be able to list "old" stored keys.

File: app/plugin_key_value_store.go (hunk 4 above)
Comment (hanzei) [reply app/plugin_key_value_store.go]: Sorry for being imprecise here. https://github.com/mattermost/mattermost-server/pull/9817/commits/d5a50afc803d0ba09268ef7b5d558011c960b811 was not what i meant. Can you please take a look at @lieut-data comment?

File: app/plugin_key_value_store.go (hunk 7 above)
Comment (lieut-data) [reply app/plugin_key_value_store.go]: `KVList` /should/ be just as useful in terms of listing the keys that can be fetched, but I agree that we lose an opportunity to "convert" back to the original key, until the plugin decides to write it again. I still think the "principle of no surprise" is best applied here, since we can't guarantee ever recovering the original keys at all, and we do still at least try to update them on set.

Comment (DSchalla) [reply app/plugin_key_value_store.go]: That makes sense, so the consensus appears to be to _not_ update the records on a get @gupsho - Sorry for the discussing as part of your PR.

Comment (gupsho) [reply app/plugin_key_value_store.go]: Let me update it.

File: app/plugin_key_value_store.go [hunk 11]
```diff
@@ -23,41 +23,47 @@ func (a *App) SetPluginKey(pluginId string, key string, value []byte) *model.App
...
 
-	if result.Err != nil {
-		if result.Err.StatusCode == http.StatusNotFound {
-			return nil, nil
-		}
+	if result.Err == nil {
+		kv := result.Data.(*model.PluginKeyValue)
+		return kv.Value, nil
+	}
+
```
Comment (hanzei) [inline app/plugin_key_value_store.go]: I think 
```go
if result.Err.StatusCode != http.StatusNotFound {
		mlog.Error(result.Err.Error())
		return nil, result.Err
}
```
is missing here.

Comment (lieut-data) [reply app/plugin_key_value_store.go]: Ah, good catch! I looked at that code, even checked it out, but convinced myself it was ok. I'll patch that on @gupsho's behalf to get this merged today.

**Comment by gupsho** [comment]
Thanks for fixing it.

**Comment by lieut-data** [comment]
@gupsho, you're welcome, and thank *you* for helping us get this into v5.6 before code complete :)

**Comment by jasonblais** [comment]
Huge thanks @gupsho for this contribution!!

**Comment by hanzei** [comment]
@gupsho If you are interested to join the discussion about plugin, feel free to join the [Mattermost Community Server](https://community.mattermost.com).
